osutil/user/user.go

```diff
@@ -0,0 +1,281 @@
+	"strconv"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil/sys"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var minUserUID sys.UserID
+var shells = []string{"/bin/sh"}
+
+func init() {
+	Init()
+}
+
+func Init() {
+	// any others?
+	if release.DistroLike("fedora") {
+		minUserUID = 500
+	} else {
+		minUserUID = 1000
+	}
+
+	if f, err := os.Open(filepath.Join(dirs.GlobalRootDir, "/etc/shells")); err == nil {
+		defer f.Close()
+		scanner := bufio.NewScanner(f)
+		for scanner.Scan() {
+			line := bytes.TrimSpace(scanner.Bytes())
+			if len(line) == 0 || line[0] == '#' {
+				continue
+			}
+			shells = append(shells, string(line))
+		}
+	}
+
+	me, meErr = FromUID(sys.Getuid())
+}
+
+// IsNonSystem is a filter that will return true only if the given
+// user is not a "system" user. The exact meaning of this is
+// backend-dependent, but in general system users aren't 'people'.
+func IsNonSystem(u *User) bool {
+	return !u.isSystemUser()
+}
+
+type User struct {
+	name  string
+	home  string
+	shell string
+	uid   sys.UserID
+	gid   sys.GroupID
+}
+
+func (u *User) Name() string {
+	return u.name
+}
+
+func (u *User) Home() string {
+	return u.home
+}
+
+func (u *User) UID() sys.UserID {
+	return u.uid
+}
+
+func (u *User) GID() sys.GroupID {
+	return u.gid
+}
+
+func (u *User) isSystemUser() bool {
+	if u.uid > 0 && u.uid < minUserUID {
+		return true
+	}
+
+	if u.name == "nobody" {
+		return true
+	}
+
+	return !strutil.ListContains(shells, u.shell)
+}
+
+var me *User
+var meErr error
+
+func Current() (*User, error) {
+	return me, meErr
+}
+
+// NotFound is returned by First when no users are found that match the given filter.
+var NotFound = errors.New("user not found")
+
+// First tries to return a user that matches the given filter.
+func First(filter func(*User) bool) (*User, error) {
+	var it Iter
+	defer it.Finish()
+	for it.Scan(filter) {
+		return it.User(), nil
+	}
+
+	err := NotFound
+	if it.err != nil {
+		err = it.err
+	}
+
+	return nil, err
+}
+
+// FromUID tries to find a user with the given UID.
+func FromUID(uid sys.UserID) (*User, error) {
+	return First(func(u *User) bool {
+		return u.uid == uid
+	})
+}
+
+// FromName tries to find a user with the given name.
+func FromName(name string) (*User, error) {
+	return First(func(u *User) bool {
+		return u.name == name
+	})
+}
+
+var passwds = []string{
+	// order could be important
+	"/var/lib/extrausers/passwd",
+	"/etc/passwd",
+}
+
+// an Iter will iterate over the user database.
+type Iter struct {
+	pwi     int
+	pwf     *os.File
+	scanner *bufio.Scanner
+	cur     *User
+	err     error
+}
+
+// User returns the user found by Scan.
+func (it *Iter) User() *User {
+	return it.cur
+}
+
+// Scan advances the iterator until it finds a user that matches all
+// the given conditions.
+func (it *Iter) Scan(conds ...func(*User) bool) bool {
+	if it.err != nil {
+		return false
+	}
+
+	for it.scanner == nil {
```

Comment:
Maybe its just because its Monday morning, but I find this for it.scanner == nil loop slightly confusing. Maybe because it requires me to understand relatively many of the state of the iterator. I was thinking maybe something more high level here, maybe something like:
```
if !it.havePwFile() {
    if ok = it.nextPwFile(); !ok { return false }
}
```
Full example: https://gist.github.com/mvo5/f5da98a50ce088acc9a70ff5a2ad4867 - I'm not too fond of the names and the recursion inside nextPwFile() but I find it easier to read once things are split out a bit.

Reply:
heh. The code used to have a separate `nextFile` method, but after looking at it for too long I guess the complexity evaporated for me. Looks like I was wrong and need to split it back out; no worries.